### DISCUSSION TITLE
Syncing library between multiple computers with Syncthing

### DISCUSSION THREAD

**Discussion by CrazyVito11** [ppy/osu#27744] [Q&A] [answered]
I personally use [Syncthing](https://syncthing.net/) to sync all kinds of things between my computers, but I wonder if I could do the same for my osu! library by simply syncing the `/.local/share/osu/files` directory.

I looked at the [User file storage wiki page](https://github.com/ppy/osu/wiki/User-file-storage) to see if anything was mentioned here about syncing or any important files that you need to watch out for, but I could draw a definitive answer from this to see if it is safe to do.

Is there anyone that could provide some insight if this would be safe to do so, or if anyone has experience with doing this exact thing? Any information is appreciated :-)

**Extra information in case it's needed:**
OS: Arch Linux, running kernel `6.8.1-arch1-1`
Syncthing version: `v1.27.4, Linux (64-bit Intel/AMD)`
Installation method: [osu-lazer-bin AUR package](https://aur.archlinux.org/packages/osu-lazer-bin) _(From checking the PKGBUILD file, it's just the official appimage release, but with a desktop file and icons added to it so it shows up like a normal application)_

**Comment by Walavouchey** [comment]
as long as `/.local/share/osu/` are the same in both computers (you also have to consider the `client.realm` file. otherwise the game won't know what's in your `files` folder), the installations should behave the same

note that the `game.ini` file contains your session token if you have it saved, meaning you don't want to leak this file

**Reply by CrazyVito11** [reply]
So just syncing the `~/.local/share/osu/files` directory and `~/.local/share/osu/client.realm` should be enough? Or is the `game.ini` also required for it to work properly?

I'll probably configure my desktop to be send only, this way it should also prevent conflicts within the `client.realm` file in case I write to both while they are unable to sync _(Syncthing is P2P, so they need to be turned on at the same time)_.

Though this does mean that scores that are made on my laptop won't be send to my desktop, but I think that is a trade off I am willing to make to prevent having to manually resolve conflicts.

**Reply by Walavouchey** [reply]
the setup you describe seems sound and should work in theory. you can try it out for a while and report back any issues you experience as it may be good to know for reference

`game.ini` only contains settings and isn't related to any other user files

**Reply by CrazyVito11** [reply]
I also noticed that the `client.realm` has a lock file called `client.realm.lock`, should I also sync that one?

**Reply by Walavouchey** [reply]
it only appears when the game is open (i.e. reading from client.realm), so likely not

**Comment by CrazyVito11** [comment, marked as answer]
# Setup
The setup I've used, consists of the following parts:

- Desktop (Main machine)
- Server (Always online machine)
- Laptop (Secondary machine)

The server is optional, but I added it anyways since that machine is always online. I rarely have my desktop and laptop turned on at the same time, so putting the server in the middle allows my laptop to always sync it's library, even if my desktop is turned off.

# Configuration
I've configured my desktop to share the `~/.local/share/osu` directory, and then shared it with my server, which then shares it to my laptop.

So syncing my laptop looks something like this
Desktop _(Send only)_ -> Server _(Send & Receive)_ -> Laptop _(Receive only)_

On each device, I've also added the following ignore pattern to only sync the files that we need.
```
// Files that we want to sync
!files
!client.realm

// Ignore everything else
*
```

You could also add `!game.ini` if you also want your settings to be synced, but I haven't tried that.

# Results
I was able to play or add a song on my desktop, close the game and then open it on my laptop for it to also appear over there.
It all seems to work fine from the small amount of time I've spend testing it.

The only issue I ran into, was that sometimes new replays didn't show up on my laptop.
**Desktop**
![Screenshot_20240328_162906](https://github.com/ppy/osu/assets/31567466/3908b26d-3eec-47bd-b666-cba8b748e792)

**Laptop**
![Screenshot_20240328_162858](https://github.com/ppy/osu/assets/31567466/91f6a194-b78d-40c1-8fbd-76b0ff91f35e)

This is either because Syncthing hasn't done a rescan yet, or because my laptop also has a local change inside `client.realm` and is blocking the sync.
Forcing my desktop to rescan immediately or telling my laptop to revert the local changes fixes that issue.

---

![Screenshot_20240328_162006](https://github.com/ppy/osu/assets/31567466/b35bcea3-67a2-4bed-bf85-d847f3fca5d6)
Of course, any songs you've added or played on the laptop, will be lost when the next sync happens due to the Receive only option I've used. You could try to set both devices to Send & Receive, but I haven't tested that and will most likely cause issues if you try to run the game on both machines at the same time, or if a sync failed at some point. 

If a conflict does happen, it will select the newest file as leading, and creates a copy of the older one, so you can pick which one you want to use manually. This does mean you will lose some data, depending on which one you pick.
![Screenshot_20240328_165508](https://github.com/ppy/osu/assets/31567466/30fb7746-3c26-4f8f-921f-d56482bf2c3e)


If I run into any new issues in the future, I will update this post.
